### ISSUE TITLE
Support ActivityPub

### ISSUE THREAD

**Issue by evanp** [Elgg/Elgg#10899]
Elgg should support the [ActivityPub](https://www.w3.org/TR/activitypub/) standard for federated social networking and social API. The standard has been in development at the W3C for 2 years. It's based on the work that I and others have done to upgrade and streamline OStatus. Other social applications like pump.io and GNU MediaGoblin are working to implement ActivityPub.

There is some more information on the state of the [ActivityPub network](https://www.w3.org/wiki/Socialwg/ActivityPub_network) on the W3C wiki. We're trying to get everyone involved in open source social networking to rally around this standard and get a truly federated network working.

**Comment by benwerd** [comment]
ActivityPub is pretty cool. We're planning on implementing it over at [Known](https://github.com/idno/Known/issues/1701).

**Comment by hypeJunction** [comment]
Elgg should many things. Sadly there are too many ideas with no actual contributions. I don't think this is going to get anyone's attention unless you put your own time into it and make a PR

**Comment by benwerd** [comment]
+1 to @hypeJunction - I think what I'll try and do is take the Known code, and see if it can be rewrapped into an Elgg plugin afterwards. That way it won't require participation from the core Elgg team, who are already overstretched.

**Comment by jdalsem** [comment]
This currently will not get the focus of anyone at the core team. Feel free to provide a PR or a plugin with the features you need so we can discuss it further then. Closing this for now as i expect no activity on this issue.

**Comment by strypey** [comment]
@benwerd 
> what I'll try and do is take the Known code, and see if it can be rewrapped into an Elgg plugin afterwards

I see that Known, like Elgg, is PHP. So are GNU Social, postActiv, Hubzilla, and Friendica, all of which are working on implementing ActivityPub (or have already rolled it out). Maybe the best long-term solution would be to create a PHP federation library that supports ActivityPub, and potentially other protocols, as Jason from SocialHome has done with Ruby (see: https://github.com/jaywink/socialhome/issues/320#issuecomment-385733161)? That way devs from all the PHP projects that want to federate can contribute to one codebase as they are able, rather than doing essentially the same work in parallel.

**Comment by hypeJunction** [comment]
I think we would have to start with support for ActivityStreams, but just as @strypey there is no decent wrapper to work with for either ActivityStreams or ActivityPub. I personally don't want to start from scratch.

**Comment by aschrijver** [comment]
Update FYI - Currently there are 2 PHP ActivityPub libraries:

- [ActivityPhp](https://landrok.github.io/activitypub/)
- [ActivityPub-PHP](https://github.com/pterotype-project/activitypub-php)

**Comment by mkljczk** [comment]
Yup @aschrijver. And the list of actual PHP software with mature ActivityPubs implementation is getting bigger, there's Friendica, Pixelfed, Hubzilla (with addon), WordPress AP plugin, NextCloud Social…

**Comment by Facyla** [comment]
May we reopen this ?  As the protocol use seems to increase, with better adoption, it might be interesting to keep track of projects (also besides core) working on this in the Elgg world. 

Note : https://landrok.github.io/activitypub/ seems maintained and might be a good candidate for tests.

**Comment by jdalsem** [comment]
I will reopen this for now. Just to keep it in the picture. If there are people willing to investigate time into integrating this in Elgg that would be nice. At the moment i do not expect core development team to put in any effort. I will check back for activity in this ticket in a year :)

**Comment by aschrijver** [comment]
Great to hear! I announced this to the fediverse: https://mastodon.social/@humanetech/107445400205978550

**Comment by RiverVanRain** [comment]
My small team are starting to do it right now.
ActivityPub is an important part of what should be in Elgg.

**Comment by jdalsem** [comment]
> My small team are starting to do it right now.

Good to hear. Make sure to submit a PR on the 4.x branch (this is the feature branch). If you make a draft PR, you can receive feedback early!

**Comment by dansup** [comment]
I'm the developer of [pixelfed](https://github.com/pixelfed/pixelfed), a php based ActivityPub implementation. Feel free to reach out if you need any assistance with implementing ActivityPub support!

**Comment by jdalsem** [comment]
Again going to close this as there is no progress. @RiverVanRain if you have some development you wish to share, please make a PR

**Comment by RiverVanRain** [comment]
I'll make it a separate plugin.
Now [IndieWeb](https://github.com/RiverVanRain/indieweb ), required for ActivityPub, is almost ready.

**Comment by wakest** [comment]
hey @RiverVanRain did did you ever start work on this ActivityPub plugin?

**Comment by RiverVanRain** [comment]
> hey @RiverVanRain did did you ever start work on this ActivityPub plugin?

@wakest Yes, we started working on the plugin on April 28th and are actively continuing to work on it.
Code will be published [here](https://github.com/RiverVanRain/activitypub).
I'll do it when the plugin is ready for testing.

Any help is welcome! 😉

**Comment by RiverVanRain** [comment]
Done https://github.com/RiverVanRain/activitypub
